programs/cross-pile/src/lib.rs

```diff
@@ -1,319 +1,767 @@
+            ctx.accounts.challenge.initiator.as_ref(),
+            &[ctx.accounts.challenge.bump]
         ];
+        anchor_spl::token::transfer(
+            CpiContext::new_with_signer(
+                ctx.accounts.token_program.to_account_info(),
+                anchor_spl::token::Transfer {
+                    from: ctx.accounts.initiator_tokens_vault.to_account_info(),
+                    to: initiator_tokens_taker,
+                    authority: ctx
+                        .accounts
+                        .challenge
+                        .to_account_info(),
+                },
+                &[&challenge_seeds[..]],
+            ),
+            ctx.accounts.challenge.initiator_wager_token_amount,
+        )?;
 
-        let signer = &[
-            &coin_seeds[..]
+        // transfer winner's tokens vault to winner's own tokens taker account
+        // ie the winner is recouping their own bet
+        anchor_spl::token::transfer(
+            CpiContext::new_with_signer(
+                ctx.accounts.token_program.to_account_info(),
+                anchor_spl::token::Transfer {
+                    from: ctx.accounts.acceptor_tokens_vault.to_account_info(),
+                    to: acceptor_tokens_taker,
+                    authority: ctx
+                        .accounts
+                        .challenge
+                        .to_account_info(),
+                },
+                &[&challenge_seeds[..]],
+            ),
+            ctx.accounts.challenge.acceptor_wager_token_amount,
+        )?;
+
+        // Close the escrow accounts
+        anchor_spl::token::close_account(
+            CpiContext::new_with_signer(
+                ctx.accounts.token_program.to_account_info(),
+                anchor_spl::token::CloseAccount {
+                    account: ctx
+                        .accounts
+                        .initiator_tokens_vault
+                        .to_account_info(),
+                    destination: ctx.accounts.initiator.to_account_info(),
+                    authority: ctx
+                        .accounts
+                        .challenge
+                        .to_account_info(),
+                },
+                &[&challenge_seeds[..]],
+        ))?;
+
+        anchor_spl::token::close_account(
+            CpiContext::new_with_signer(
+                ctx.accounts.token_program.to_account_info(),
+                anchor_spl::token::CloseAccount {
+                    account: ctx
+                        .accounts
+                        .acceptor_tokens_vault
+                        .to_account_info(),
+                    destination: ctx.accounts.acceptor.to_account_info(),
+                    authority: ctx
+                        .accounts
+                        .challenge
+                        .to_account_info(),
+                },
+                &[&challenge_seeds[..]],
+        ))?;
+
+        Ok(())
+    }
+
+    pub fn cancel_before_acceptor<'info>(
```

Comment:
Originally, this was a single method (cancel before acceptor + cancel after acceptor), but I was unable to load the acceptor's receiving token account via remaining accounts with the correct owner. The owner was showing as the token program itself, but this is a pretty important check to make sure the acceptor receives their bet back.